### DISCUSSION TITLE
Multi-file editing, code review, custom instructions, and more for GitHub Copilot

### DISCUSSION THREAD

**Discussion by Akash1134** [community/community#142982] [Copilot News and Announcements]
![Screenshot 2024-10-29 at 10 56 46 PM](https://github.com/user-attachments/assets/d5c64592-d414-40a1-aa3c-ea80e491f5c1)

Hey devs, we’ve got some powerful new upgrades for GitHub Copilot Chat in the latest Visual Studio Code release—features designed to streamline coding, debugging, and testing. Excited? So are we!

**Setting:** _github.copilot.chat.edits.enabled_

In this update, *multi-file editing* (currently in preview) lets you run AI-assisted edits across multiple files, all in one session. This means you can ask Copilot to suggest and apply code changes across files in your workspace, review in context, and refine directly in the editor. 

<img width="711" alt="Screenshot 2024-10-29 at 16 57 46" src="https://github.com/user-attachments/assets/3f3a708a-c5d7-473d-ba31-580f30dfa9f1">

This feature is ideal for tackling larger code changes that span multiple files—keeping the Copilot Chat flow right beside your code, while maintaining fast feedback in the editor. It’s a blend of the conversational ease of Copilot Chat and the precision of in-line coding.

We’d love to hear from you: what coding tasks are you hoping to streamline with multi-file editing? Or, if you've tried it, what’s been your favorite use so far? Let’s build the future of GitHub Copilot together! 

Get more details about [multi-file editing](https://code.visualstudio.com/docs/copilot/copilot-edits) in the VS Code documentation.

### [Read more here](https://github.blog/changelog/2024-10-29-multi-file-editing-code-review-custom-instructions-and-more-for-github-copilot-in-vs-code-october-release-v0-22/)

**Comment by ianshulx** [comment]
That sounds exciting! The new multi-file editing feature in GitHub Copilot Chat for Visual Studio Code is a game-changer. Being able to make AI-assisted code changes across multiple files in one session will definitely streamline the coding process. It's great to see how it combines the conversational capabilities of Copilot Chat with direct in-line coding. I'm looking forward to trying it out and seeing how it enhances the workflow for larger code changes!

**Comment by Nathan-Bransby-NMT** [comment]
Absolutely amazing stuff! :octocat:

**Comment by Marcono1234** [comment]
The screenshot in the [changelog post for code review](https://github.blog/changelog/2024-10-29-github-copilot-code-review-in-visual-studio-code-public-preview/) seems to show an incorrect Copilot suggestion:
The message says "Should be 'req.params'", but then in the suggested change it actually replaced it with `req.body`.

Might want to use one of the screenshots from the documentation instead, where the suggested changes seem to be correct.

Besides that this feature seems quite interesting though.

Edit: Looks like the screenshot has been replaced now.

**Comment by nnmrts** [comment]
I don't know what I'm doing wrong but I can't seem to access this feature at all...

![Bildschirmfoto 2024-10-31 um 08 01 29](https://github.com/user-attachments/assets/34bfcfc8-a98a-47c7-b157-a281cebe20ca)

Version: 1.95.0-insider
Commit: 4b4bf8f594077c3dc53ef70518531f64ed566194
Date: 2024-10-11T18:44:53.770Z (2 wks ago)
Electron: 32.1.2
ElectronBuildId: 10323173
Chromium: 128.0.6613.162
Node.js: 20.17.0
V8: 12.8.374.33-electron.0
OS: Darwin arm64 24.0.0

**Comment by nnmrts** [comment]
Ah, nevermind, seems like the "searching for updates" was broken on 1.95.0 insiders.

**Reply by samus-aran** [reply]
? Care to provide more info? Has it be resolved on 1.95.0 insiders? Is there a reference to an issue?

**Reply by nnmrts** [reply]
Not at all, no. At least I didn't find anything. I just had to download the latest binary manually.

All I can say is, I pressed "check for updates" multiple times and nothing happened and also it never auto-updated in the last two weeks.

**Comment by darkostanimirovic** [comment]
I think it's a great step forward. However, what if I'm not sure exactly which files need changing? I expect to be able to say "add functionality X" and Copilot figures out which files should be affected. Is that on the roadmap?

**Comment by patricknelson** [comment]
This is awesome! Been looking forward to multi-file editing for a long time, especially in JetBrains IDEs. Is there a roadmap for when that might become available?

**Reply by joshualyon** [reply]
I'm curious about this as well - I would love to see the GitHub Copilot plugin for JetBrains support Multi-file Editing. 

I gave the Multi-file Editing a try in VS Code and it's a great experience... it's really useful to be able to use the `#FileName` references or <kbd>⌘</kbd> + <kbd>/</kbd> to quickly pull in the relevant files for context and make edits across multiple related files (eg. a class definition and various uses of that class). And it's great to be able to iterate on those changes and see the inline diffs to better understand what changed and make sure existing code doesn't get whacked as part of the edits. 

PS. I found the diffs from Multi-file Editing in VS Code a bit quirky as sometimes the diffs wouldn't display correctly, but usually just switching away from the file and back would resolve it. Anyway, **would love to see Multi-file editing in Jetbrains** with Copilot!

**Comment by heheda123123** [comment]
Copilot Edits is very easy to use, when will it be integrated into Visual Studio?
